File: day17/src/main/cpp/solution.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <regex>
#include <string>
#include <vector>

using namespace std;
#define coordinate pair<int, int>

pair<coordinate, coordinate> readInputFile() {
  ifstream file("../resources/input");
// ifstream file("../../test/resources/part1/example1.input");
  string s;
  smatch match;
  regex re("target area: x=(-?\\d+)..(-?\\d+), y=(-?\\d+)..(-?\\d+)");
  getline(file, s);
  regex_search(s, match, re);
  return pair(coordinate(stoi(match.str(1)), stoi(match.str(2))),
              coordinate(stoi(match.str(3)), stoi(match.str(4))));
}

vector<int> simulateTrajectoryHeights(coordinate xRange, coordinate yRange) {
  vector<int> result;
  for (int x = -xRange.second; x <= xRange.second; ++x) {
    for (int y = yRange.first; y <= -yRange.first; ++y) {
      coordinate current(0, 0);
      int xDiff = x, yDiff = y, maxY = INT32_MIN;
      while (current.first + xDiff <= xRange.second &&
             current.second + yDiff >= yRange.first) {
        current = coordinate(current.first + xDiff, current.second + yDiff);
        maxY = max(maxY, current.second);
        yDiff--;
        if (xDiff != 0) {
          xDiff += xDiff > 0 ? -1 : 1;
        }
      }
      if (current.first >= xRange.first && current.first <= xRange.second &&
          current.second >= yRange.first && current.second <= yRange.second) {
            //   cout << x << " " << y << " " << current.first << " "<< current.second << endl;
        result.push_back(maxY);
        
      }
    }
  }

  return result;
}

int main() {
  auto input = readInputFile();
  auto heights = simulateTrajectoryHeights(input.first, input.second);
  cout << "part 1" << endl
       <<  *max_element(heights.begin(), heights.end())
       << endl
       << "part 2" << endl
       << heights.size() << endl;
  return 0;
}

File: day17/src/main/python/solution.py
import re


def read_input():
    f = open('../resources/input', 'r')
    line = f.readline().strip()
    result = re.search(r'target area: x=(-?\d+)..(-?\d+), y=(-?\d+)..(-?\d+)', line)
    return (int(result.group(1)), int(result.group(2))), (int(result.group(3)), int(result.group(4)))


def simulate_trajectories(xRange, yRange):
    results = []
    for x in range(-xRange[1], xRange[1]+1):
        for y in range(yRange[0], -yRange[0]+1):
            coordinate = (0, 0)
            xValue = x
            yValue = y
            maxY = -999999
            while coordinate[0] + xValue <= xRange[1] and coordinate[1] + yValue >= yRange[0]:
                coordinate = (coordinate[0] + xValue, coordinate[1] + yValue)
                maxY = max(maxY, coordinate[1])
                yValue -= 1
                if xValue != 0:
                    xValue += 1 if xValue < 0 else -1
            if xRange[0] <= coordinate[0] <= xRange[1] and yRange[0] <= coordinate[1] <= yRange[1]:
                results.append(maxY)
    return results


if __name__ == "__main__":
    x_range, y_range = read_input()
    trajectories = simulate_trajectories(x_range, y_range)
    print("part 1")
    print(max(trajectories))
    print("part 2")
    print(len(trajectories))
